### PR TITLE
[36233] OpenProject Dark Mode

### PR DESCRIPTION
**Pull request by bsatarnejad**
https://community.openproject.org/projects/openproject/work_packages/36233/activity

### CODE REVIEW

**Comment by lindenthal** [comment]
Hi @HDinger,
How do I activate the feature flag on PP?
Best
Niels

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: That variable does not exist any more.

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: `--color-btn-border` does also not exist any more. It is now named `--button-default-borderColor-rest`

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: It looks like this doesn't work correctly:
Bildschirmfoto 2024-04-26 um 08 40 09

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: That variable also does not exist any more.

File: frontend/src/global_styles/content/_pagination.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_pagination.sass:83]: `--button--background-color` is also deprecated by now

File: frontend/src/global_styles/openproject/_scm.sass
Comment (HDinger) [inline frontend/src/global_styles/openproject/_scm.sass]: Please undo this change as `fgColor-muted` is the correct variable name

File: frontend/src/global_styles/common/select/select.sass
Comment (HDinger) [inline frontend/src/global_styles/common/select/select.sass]: `--list-item-hover--color` does not exist any more

**Review by HDinger — COMMENTED** [review]
Hi @bsatarnejad 

I know that this is still in a Draft state so I did not check the code into detail yet. I just found some places where I originally used the old (deprecated) variables which are now not supported any more.

File: frontend/src/global_styles/content/_pagination.sass
Comment (bsatarnejad) [reply frontend/src/global_styles/content/_pagination.sass:83]: Hi! --button--background-color is one of our global variables, it's not deprecated, is it?

File: app/helpers/application_helper.rb
Comment (HDinger) [inline app/helpers/application_helper.rb]: Since the if statement is inside the brackets, these brackets will always be added as an option, resulting in an empty object being shown if the fetaure flag is deactivated
Bildschirmfoto 2024-06-12 um 07 22 32

Bildschirmfoto 2024-06-12 um 07 22 25

You need to change that a bit to something like this:

```
  def theme_options_for_select
    options = [
      [t("themes.light"), "light"],
      [t("themes.light_high_contrast"), "light_high_contrast"]
    ]
    options.push([t("themes.dark"), "dark"]) if OpenProject::FeatureDecisions.dark_mode_active?
    options
  end
```

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: Why did you remove this block? We definetly need that to enforce that the Primer buttons have our custom color.
Bildschirmfoto 2024-06-12 um 07 52 57

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: I think, this extra selector is not needed as `[data-dark-theme=dark]` is currently always matching when `:not([data-light-theme=light_high_contrast]` is already matching

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: I think you have to cleanup this whole file a bit. It is kind of hard to understand when which rules are applied. Ideally, this file would have only three rule blocks (like it was before)

1. Generic stlyes that should always be active (`[data-color-mode]`)
2. Specific styles for the high_contrast mode which are overriding stuff, e.g. the header or the main menu color (`[data-light-theme=light_high_contrast]`)
3. Styles that are only valid if we are **not** in the accessibility mode (`[data-color-mode]:not([data-light-theme=light_high_contrast]) `) 

I don't really see a reason for styles that are only applied in the dark mode, but maybe you know a case.

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb]: I think that the dark mode should be using the configured color variables for accent and primary actions like we do in the normal light mode. But we may have to double check with the product team

File: frontend/src/app/features/in-app-notifications/entry/in-app-notification-entry.component.sass
Comment (HDinger) [inline frontend/src/app/features/in-app-notifications/entry/in-app-notification-entry.component.sass]: Both variables look semantically wrong especially since they are both refering to different things but are used for the same thing within OpenProject. Please double check those. GitHub also has a notification center, maybe you find a better suited variable there.

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass
Comment (HDinger) [inline frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass:21]: The muted variants are really hard to see. I think a stronger color would be better for both cases

Bildschirmfoto 2024-06-12 um 08 33 01

File: frontend/src/global_styles/common/header/app-menu.sass
Comment (HDinger) [inline frontend/src/global_styles/common/header/app-menu.sass]: I think this should be the body-background color and the overlay background

File: frontend/src/global_styles/common/menu/menu.sass
Comment (HDinger) [inline frontend/src/global_styles/common/menu/menu.sass:29]: This color is semantically wrong as there is no control element

File: frontend/src/global_styles/common/menu/menu.sass
Comment (HDinger) [inline frontend/src/global_styles/common/menu/menu.sass:34]: Same as above, please try to find a semantically fitting color

File: frontend/src/global_styles/common/select/select.sass
Comment (HDinger) [inline frontend/src/global_styles/common/select/select.sass:20]: Same here, please do not use `control` specific styles for non-control elements

File: frontend/src/global_styles/content/_accounts.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_accounts.sass]: Also semantically not really fitting as there is no Overlay

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (HDinger) [inline app/views/custom_styles/_primer_color_mapping.erb:24]: I know it was like this before, but that feels also a bit weird.

File: frontend/src/global_styles/content/_autocomplete.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_autocomplete.sass:151]: I saw that pattern quite often now of using the control colours for something that is no control element. Could you please check whether there is an alternative?

File: frontend/src/global_styles/content/_datepicker.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_datepicker.sass:135]: This is indeed much darker in the normal mode now. That might be okay, we have to check with product or find an alternative

File: frontend/src/global_styles/content/_datepicker.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_datepicker.sass:1]: For me the datepicker looks fine, although there are some differences. I think we have to talk with product about that.

File: frontend/src/global_styles/content/_pagination.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_pagination.sass]: These are not the correct colors as it looks off in the normal and dark mode

Bildschirmfoto 2024-06-12 um 07 17 05
Bildschirmfoto 2024-06-12 um 07 35 01

File: frontend/src/global_styles/content/_pagination.sass
Comment (HDinger) [reply frontend/src/global_styles/content/_pagination.sass:83]: Yes, your are right. My bad 🙈

File: frontend/src/global_styles/content/_toast.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_toast.sass]: The muted styles are semi transparent which looks weird on the notifications
Bildschirmfoto 2024-06-12 um 07 22 56

File: frontend/src/global_styles/content/_tooltip.sass
Comment (HDinger) [inline frontend/src/global_styles/content/_tooltip.sass]: Is that the correct variable?

File: frontend/src/global_styles/openproject/_mixins.sass
Comment (HDinger) [inline frontend/src/global_styles/openproject/_mixins.sass]: That looks also semantically incorrect. Github has widget boxes as well. e.g for pinned repositories. Please double check what they use there

File: frontend/src/global_styles/openproject/_variable_defaults.scss
Comment (HDinger) [inline frontend/src/global_styles/openproject/_variable_defaults.scss]: Why should the task board header remain white?

**Review by HDinger — CHANGES_REQUESTED** [review]
Hi @bsatarnejad 

Great job, it already looks really nice 🥳 👍  No worries, I have made a lot of comments but they are all rather small. In many cases I was just unsure whether the used variable is the correct one. Please double check those and address the remarks I made in code. 

Additionally, I found those things that we should address before merging:

- [x] [Normal mode] There is a border missing around the toggle in the global project dropdown
Bildschirmfoto 2024-06-12 um 07 17 22

- [x] [ckEdior] There are some issues in the ckEditor, when hovering or opening dropdowns
 
Bildschirmfoto 2024-06-12 um 07 28 46

Bildschirmfoto 2024-06-12 um 07 28 51

Bildschirmfoto 2024-06-12 um 07 29 04

 - [x] [ckEdior/Wiki] When creating special elements, like a table of contents, a header row/column in a table or a code block, they have the wrong color
Bildschirmfoto 2024-06-12 um 07 29 23

Bildschirmfoto 2024-06-12 um 07 31 24


- [x] [Gantt] The information bubbles on the right and left side of a gantt bar are light grey
Bildschirmfoto 2024-06-12 um 07 35 24


- [x] [WP table] Hovering a disabled field (e.g. updated on) creates a white background
Bildschirmfoto 2024-06-12 um 07 36 38

- [x] [WP full view] Opening a multi-select field, there are white controls
Bildschirmfoto 2024-06-12 um 07 37 16


-------------------
Further, I found some additional things that needs fixing but can be done in a second PR:

- [x] The bubble indicator in buttons has the wrong color
Bildschirmfoto 2024-06-12 um 07 33 54
Bildschirmfoto 2024-06-12 um 07 34 36


- [x] Loading indicators are based on white

Bildschirmfoto 2024-06-12 um 07 34 16

Bildschirmfoto 2024-06-12 um 08 25 13


- [x] Hovering a board card creates a white fading effect. Further the cards have a white shadow
Bildschirmfoto 2024-06-12 um 07 33 45


- [ ] Disabled select fields look to bright
Bildschirmfoto 2024-06-12 um 07 27 24

- [ ] The select boxes seem to have a darker background than normal input fields
Bildschirmfoto 2024-06-12 um 07 25 12


- [ ] On the overview page, the graph text is too dark to read
Bildschirmfoto 2024-06-12 um 07 24 37

- [ ] The input on the repository page and the cost reports page has the wrong color
Bildschirmfoto 2024-06-12 um 07 31 40
Bildschirmfoto 2024-06-12 um 07 32 11

- [x] The progress bar on the Roadmap is white
Bildschirmfoto 2024-06-12 um 07 36 17


- [ ] Last but not least, there is the known issue of using typ
Bildschirmfoto 2024-06-12 um 07 35 04
es / status with a dark color

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (bsatarnejad) [reply app/views/custom_styles/_primer_color_mapping.erb]: Hi @HDinger ! Sorry, I thought that like high-contrast mode, we want to use primer variables for primary buttons for dark mode. I revert it.

File: app/views/custom_styles/_primer_color_mapping.erb
Comment (bsatarnejad) [reply app/views/custom_styles/_primer_color_mapping.erb]: I asked Parimal about it and am waiting for his answer, but for now I revert the code, then we only use primer variables for our accent and primary colors only in high contrast mode.

File: frontend/src/global_styles/openproject/_variable_defaults.scss
Comment (bsatarnejad) [reply frontend/src/global_styles/openproject/_variable_defaults.scss]: can we talk about task board colors with product team?

File: frontend/src/global_styles/content/_toast.sass
Comment (bsatarnejad) [reply frontend/src/global_styles/content/_toast.sass]: yeah I agree. I checked the Primer Design System, they are using muted variables for banners. I used the same variables as they used, and it works well on high-contrast and light modes, but for dark mode it's a bit weird. Do we want to use primer variables for toasts?

Comment (HDinger) [reply frontend/src/global_styles/content/_toast.sass]: What does it mean "it's a bit weird"?

Comment (bsatarnejad) [reply frontend/src/global_styles/content/_toast.sass]: I mean, with a transparent background, elements behind the notification are visible, so it is weird for users.

So for now, I will revert the changes for colors and will talk with products to see which primer color variable is suitable for toasts.

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass
Comment (HDinger) [reply frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass:21]: Ah I see, this is the same issue as we have with the normal Banners. Let's discuss that on Monday 👍

**Review by HDinger — APPROVED** [review]
Great job @bsatarnejad 👍 🥳

**Pull request merged**
